File: domain/use_case/generate_completion_use_case.py
from domain.boundary.llm_repository import LLMRepository
from domain.models.model_errors import LLMUnavailableError
from domain.models.response_model import FunctionCallResponseModel, ResponseModel
from utils.logger import Logger


class GenerateCompletionUseCase:
    def __init__(self, repository: LLMRepository, logger: Logger):
        self._repository = repository
        self._logger = logger

    def execute(self, user_input: str) -> ResponseModel:
        self._logger.info("Processing user input.", context={"length": len(user_input)})
        # TODO: Podemos aplicar nossas GuardRails aqui. Logo faremos isso.
        try:
            completion = self._repository.complete(user_input)
            self._logger.info(
                "Completion generation successful.",
                context={
                    "is_function_call": {
                        isinstance(completion, FunctionCallResponseModel)
                    }
                },
            )
            return completion
        except LLMUnavailableError as e:
            self._logger.error("LLM service unavailable.", context={"error": str(e)})
            raise e

File: src/domain/use-case/generate-completion-use-case.ts
/**
 * Use case for generating LLM completions.
 */

import type { LLMRepository } from '../boundary/llm-repository.js';
import type { ResponseModel } from '../models/response.js';
import { LLMUnavailableError } from '../models/errors.js';
import type { Logger } from '../../utils/logger.js';
import { isFunctionCallResponse } from '../models/response.js';

export class GenerateCompletionUseCase {
  constructor(
    private readonly repository: LLMRepository,
    private readonly logger: Logger
  ) {}

  async execute(userInput: string): Promise<ResponseModel> {
    this.logger.info('Processing user input.', { length: userInput.length });

    try {
      const completion = await this.repository.complete(userInput);

      this.logger.info('Completion generation successful.', {
        isFunctionCall: isFunctionCallResponse(completion),
      });

      return completion;
    } catch (error) {
      if (error instanceof LLMUnavailableError) {
        this.logger.error('LLM service unavailable.', { error: (error as Error).message });
        throw error;
      }
      throw error;
    }
  }
}
